feat: [TR-724] note model validations

ValentinPg opened the pull request:
## JIRA Card

https://widergy.atlassian.net/browse/TR-724

## Summary

- Se se valida que el note_type de la Note solo pueda ser review o critique.
 
- Se desarrolla un método `word_count` en el modelo Note que devuelve la cantidad de palabras que tiene el atributo `content`
 
- Se desarrolla en el modelo Note un método `content_length` que devuelve un String que puede tomar los valores `short`, `medium` o `long` en función de el largo del `content`. El criterio para definir este largo se encuentra definido dentro de cada Utility a través de un método llamado `content_length_criteria`.

- Se agrega una validación al modelo Note que valida que no se puedan crear reseñas que no sean `short`. 

- Se agregan, en el archivo `seeds.rb` varias, varias notes para cada usuario, para tener de prueba

- Se implementa traducciones de mensajes de error con el módulo I18n

## Screenshots

Las reviews no pueden ser largas ni medianas
<img width="694" alt="image" src="https://github.com/user-attachments/assets/3c50e9ce-9553-4dc5-82ad-8789aefaf787" />

Podemos crear notas, ver la cantidad de palabras del `content` y entiende el mensaje `content_length`
<img width="691" alt="image" src="https://github.com/user-attachments/assets/6db7eef5-1625-4ace-b1c0-924cec2063d6" />

Las notas solo pueden ser `review` o `critique`
<img width="685" alt="image" src="https://github.com/user-attachments/assets/9ebe4b49-9ca7-44ad-9a2c-0596b4027c91" />


Varios mensajes se encuentran internacionalizados
<img width="688" alt="image" src="https://github.com/user-attachments/assets/ea5fc239-c262-47be-815e-61acc1fd96a1" />

Los tests corren
<img width="564" alt="image" src="https://github.com/user-attachments/assets/519206cc-bd26-4467-85d1-9b5cbc425e11" />

## Test Cases

- Probar que no se puedan crear reviews de otro tipo que no sean `short` y probar crear `critiques` de los 3 tipos.
 
- Probar que no se puedan crear Notes que no sean del tipo `review` o `critique`.

- Probar que las Notes se generen correctamente con el archivo `seeds.rb`.

Commit:
Se cambia el tipo de la columna en la base de datos y se agrega el enum en el modelo Note

Commit:
Se agrega el metodo word_length y el correspondiente test al modelo Note

Commit:
se agregan las condiciones para short, medium y long en la utility

Commit:
se cambian las condiciones de la south utility

Commit:
Cambio en la logica de como se mide el content length en la utility

Commit:
Se separan los contenidos por long, short y medium y se limitan las reviews

Commit:
Se crean las notes en el seed

Commit:
I18n funcionando

Commit:
Dar formato al documento

ValentinPg committed:
fix: [TR-722] user serializer

mgonidev commented on config/locales/en.yml:
```diff
@@ -29,5 +29,12 @@
+
 en:
-  hello: "Hello world"
+  shorter_review: 'Reviews must be shorter'
```
Ojo que falta el salto de linea requerido acá. Eso quiere decir el símbolo que te marca GitHub

mgonidev replied:
Se que dije de hacerlo mas corto, pero me arrepentí je.
Hagamos que quede con la estructura que usaste en el es.yml

```
es:
  activerecord: 
    errors:
      models:
        note:
          shorter_review: 'La review debe ser más corta'

```

mgonidev commented on config/locales/es.yml:
```diff
@@ -30,4 +30,11 @@
 
 es:
-  hello: "Hola Mundo!"
+  activerecord: 
+    errors:
+      models:
+        note:
+          attributes:
+            note_type:
+              shorter_review: 'La review debe ser más corta'
```
mucha iteración, podríamos hacerlo mas general no? Por otra parte, donde lo estamos usando?

ValentinPg replied:
Agregado!

mgonidev commented on spec/models/note_spec.rb:
```diff
@@ -10,15 +10,16 @@
   end
 
-
   %i[title content].each do |value|
     it { is_expected.to validate_presence_of(value) }
   end
```
habiamos dicho algo con respecto a esto no?

ValentinPg replied:
Listo, separado y saqué el test de `content_length` ya que lo estoy armando para sumar en el próximo PR

mgonidev reviewed (commented):
Revisemos los coments y vamos 👍 
Revisemos los conflictos y sumemos una captura de los tests que hayan corrido ok ⚠️

ValentinPg committed:
feat: [TR-723] Note model

* Se creó el modelo de Note, la tabla Notes y se la vinculó con usuarios.

Adeás se corrieron las migraciones para aplicar estos cambios

* Se agrega la validación del note_type al modelo note.

También se le agrega al modelo de user el dependent: :destroy para las notes relacionadas

* Ya se pueden crear las notas y quedan asociadas a los usuario

* Se agregan los tests para el modelo Note

* Agrego el on update en la migración de Note

* Corrección de los cambios que me fueron pedidos

---------

Co-authored-by: ValentinPg <valentinpugliese@outlook.com>

Commit:
se agrega el salto de linea en la última linea

Commit:
se reduce la profundidad de :shorter_review en el archivo es.yml

Commit:
Merge remote-tracking branch 'origin/master' into TR-724

Commit:
cambios en los tests y migraciones

Commit:
recomendaciones de joaco sobre nombres de metodos y private

Commit:
Merge branch 'TR-723' into TR-724

mgonidev commented on config/locales/en.yml:
```diff
@@ -29,5 +29,12 @@
 # To learn more, please read the Rails Internationalization guide
 # available at http://guides.rubyonrails.org/i18n.html.
 
+# en:
+#   activerecord: 
+#     errors:
+#       models:
+#         user:
+#           shorter_review : 'Prueba'
+
```
por que como comentario?

ValentinPg replied:
Perdoná me olvidé de borrarlo

mgonidev commented on config/locales/es.yml:
```diff
@@ -30,4 +30,9 @@
 
 es:
-  hello: "Hola Mundo!"
+  activerecord: 
+    errors:
+      models:
+        note:
+          shorter_review: 'La review debe ser más corta'
+
```
ambos ym tienen que tener la misma estructura, sino falla cuando vas a buscar el mensaje.

ValentinPg replied:
Genial, ahora ya entiendo mejor como funciona esta gema

mgonidev reviewed (approved):
Te dejo el approve pero modifiquemos lo de los yml 🚀

Commit:
Correcciiones que me pidió Mati en el yml

florfrioli commented on app/models/note.rb:
```diff
@@ -5,21 +5,31 @@
+  enum note_type: { "review": 0, "critique": 1 }
   validates :title, :content, :note_type, presence: true
-  validate :note_type_validation
+  validate :review_cap, if: -> { utility.present? && content.present? }
```
Hacía falta preguntar si utility present? 🤔

florfrioli commented on db/seeds.rb:
```diff
@@ -32,3 +32,12 @@
 end
+
+# Notes
+User.all.find_each do |user|
```
Acá no está corriendo el linter pero ojo con las tabulaciones 🤗

florfrioli commented on spec/models/note_spec.rb:
```diff
@@ -6,8 +6,7 @@
   end
 
   it 'note_type must be review or critique' do
```
Cambiamos del caso feliz al triste, deberían estar ambos.

Hay unos métodos igual que los expect de abajo para los enums 🤗
